backend/gateway/mux/util.go

```diff
@@ -0,0 +1,38 @@
+package mux
+
+import (
+	"net/http"
+	"reflect"
+	"strings"
+)
+
+// All modern browsers send `text/html` as the first MIME type in the Accept header for navigation requests.
+// This is naive and doesn't do any actual content negotiation, just checks if the client accepts text/html which
+// is good enough for our use case of determining whether the request originates from the SPA's XHR network client or
+// browser navigation.
+// https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation/List_of_default_Accept_values
+func isBrowser(h http.Header) bool {
+	directives := strings.Split(h.Get("Accept"), ",")
+	for _, d := range directives {
+		mt := strings.SplitN(strings.TrimSpace(d), ";", 1)
+		if len(mt) > 0 && mt[0] == "text/html" {
+			return true
+		}
+	}
+	return false
+}
+
+// This is a hack, but it's the best way to get the Accept header from the response forwarder without significant plumbing
+// complexity and overhead via middleware. Do not invoke this in the normal request flow.
+func copyRequestHeadersFromResponseWriter(w http.ResponseWriter) http.Header {
+	req := reflect.ValueOf(w).Elem().FieldByName("req").Elem().FieldByName("Header")
+
+	h := make(http.Header, req.Len())
+	iter := req.MapRange()
+	for iter.Next() {
+		k := iter.Key().String()
+		v := iter.Value().Index(0).String()
```

Comment:
is there a chance that `Index(0)` could fail here if the value is empty?

Reply:
Technically no, headers will always have values (even if it's just `""`) but I added a check anyways.